src/Adapter/MSTest.TestAdapter/Execution/UnitTestRunner.cs

```diff
@@ -122,13 +123,19 @@ internal FixtureTestResult GetFixtureTestResult(TestMethod testMethod, string fi
         static UnitTestOutcome GetOutcome(Exception? exception) => exception == null ? UnitTestOutcome.Passed : UnitTestOutcome.Failed;
     }
 
+    // Task cannot cross app domains.
+    // For now, TestExecutionManager will call this sync method which is hacky.
+    // If we removed AppDomains in v4, we should use the async method and remove this one.
+    internal UnitTestResult[] RunSingleTest(TestMethod testMethod, IDictionary<string, object?> testContextProperties, IMessageLogger messageLogger)
+        => RunSingleTestAsync(testMethod, testContextProperties, messageLogger).GetAwaiter().GetResult();
```

Comment:
Could we maybe instead have our custom task-like type that's serializable and can cross app domains?

Reply:
Will keep it simplified for now, and we will adjust in future if needed.